ambry-store/src/main/java/com/github/ambry/store/DiskManager.java

```diff
@@ -570,4 +590,165 @@ private void reportUnrecognizedDirs() {
+
+  public DiskHealthStatus getDiskHealthStatus() {
+    return diskHealthStatus;
+  }
+
+  /**
+   Manages the disk healthchecking tasks such as creating,writing, reading, and deleting a file
+   */
+  private class DiskHealthCheck {
+    private final int operationTimeoutSeconds;      //how long a read/write operation is allotted
+    private final boolean enabled;                  //enabled healthchecking
+    private final String healthcheckDir;            //filepath of where the file will be created
+
+    DiskHealthCheck(int operationTimeoutSeconds, boolean enabled, String healthcheckDir) {
+      this.enabled = enabled;
+      this.healthcheckDir = healthcheckDir;
+      this.operationTimeoutSeconds = operationTimeoutSeconds;
+    }
+
+    /**
+     * Getter utility for enabled boolean
+     * @return flag to indicate disk healthchecking or not
+     */
+    private boolean isEnabled() {
+      return enabled;
+    }
+
+    /**
+     * Attempts to close the filechannel and delete the file
+     * @param fileChannel the file used for disk healthcheck
+     * @return True/False depending on the close/delete was successful
+     */
+    private boolean deleteFile(AsynchronousFileChannel fileChannel, boolean successfulThusFar) {
+      if (fileChannel.isOpen()) {
+        try {
+          fileChannel.close();
+        } catch (IOException e) {
+          //ensures that any error code won't be overwritten thus far
+          if (successfulThusFar) {
+            diskHealthStatus = diskHealthStatus.DELETE_IO_EXCEPTION;
+          }
+          logger.error("Exception occurred when deleting a file for disk healthcheck", e);
+          return false;
+        }
+      }
+      return true;
+    }
+
+    /**
+     * Will perform either a read or write operation on the disk while checking for exceptions and canceling the
+     * operation if there was an exception
+     * @param fileChannel used to perform the read/write operation
+     * @param buffer either used to write from or read to the buffer in the respective operation
+     * @param timeoutException indicate if the timeout is from reading or writing
+     * @param otherException indicate if the other exceptions are from reading or writing
+     * @param isWriting inform the function to either write or read with the buffer and filechannel
+     * @return True/False to indicate if the operation completed with no issues or not
+     */
+    private boolean performOperation(AsynchronousFileChannel fileChannel, ByteBuffer buffer,
+        DiskHealthStatus timeoutException, DiskHealthStatus otherException, boolean isWriting) {
+      Future<Integer> result = null;
+      try {
+        if (isWriting) {
+          result = fileChannel.write(buffer, 0);
+        } else {
+          result = fileChannel.read(buffer, 0);
+        }
+        result.get(operationTimeoutSeconds, TimeUnit.SECONDS);
+        return true;
+      } catch (TimeoutException e) {
+        diskHealthStatus = timeoutException;
```

Comment:
nit: rename this to timeoutStatus and otherStatus.

Reply:
Changed. Thanks